/*
* @license Apache-2.0
*
* Copyright (c) 2023 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

/// <reference types="@stdlib/types"/>

import srotg = require( './index' );


// TESTS //

// The function returns a collection...
{
	srotg( 0.0, 2.0 ); // $ExpectType Float32Array
}

// The compiler throws an error if the function is provided an argument which is not a number...
{
	srotg( true, false ); // $ExpectError
	srotg( false, true ); // $ExpectError
	srotg( null, null ); // $ExpectError
	srotg( undefined, undefined ); // $ExpectError
	srotg( '5', '9' ); // $ExpectError
	srotg( [], [] ); // $ExpectError
	srotg( {}, {} ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported numbder of arguments...
{
	srotg(); // $ExpectError
	srotg( 1.0, 2.0, 3.0 ); // $ExpectError
}

// Attached to the main export is an `assign` method which returns a Float32Array containing numbers...
{	
	const out = new Float32Array( 4 );

	srotg.assign( 0.0, 2.0, out, 1, 0 ); // $ExpectType Float32Array
}

// The compiler throws an error if the `assign` method is provided a first or second argument which is not a number...
{
	const out = new Float32Array( 4 );

	srotg.assign( true, true, out, 1, 0 ); // $ExpectError
	srotg.assign( false, false, out, 1, 0 ); // $ExpectError
	srotg.assign( '5', '9', out, 1, 0 ); // $ExpectError
	srotg.assign( null, null, out, 1, 0 ); // $ExpectError
	srotg.assign( [], [], out, 1, 0 ); // $ExpectError
	srotg.assign( {}, {}, out, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the `assign` method is provided a third argument which is not an array-like object...
{
	srotg.assign( 1.0, 2.0, 1, 1, 0 ); // $ExpectError
	srotg.assign( 1.0, 2.0, true, 1, 0 ); // $ExpectError
	srotg.assign( 1.0, 2.0, false, 1, 0 ); // $ExpectError
	srotg.assign( 1.0, 2.0, null, 1, 0 ); // $ExpectError
	srotg.assign( 1.0, 2.0, {}, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the `assign` method is provided a fourth argument which is not a number...
{
	const out = [ 0.0, 0.0, 0.0, 0.0 ];

	srotg.assign( 1.0, 2.0, out, '5', 0 ); // $ExpectError
	srotg.assign( 1.0, 2.0, out, true, 0 ); // $ExpectError
	srotg.assign( 1.0, 2.0, out, false, 0 ); // $ExpectError
	srotg.assign( 1.0, 2.0, out, null, 0 ); // $ExpectError
	srotg.assign( 1.0, 2.0, out, [], 0 ); // $ExpectError
	srotg.assign( 1.0, 2.0, out, {}, 0 ); // $ExpectError
	srotg.assign( 1.0, 2.0, out, ( x: number ): number => x, 0 ); // $ExpectError
}

// The compiler throws an error if the `assign` method is provided a fifth argument which is not a number...
{
	const out = [ 0.0, 0.0, 0.0, 0.0 ];

	srotg.assign( 1.0, 2.0, out, 1, '5' ); // $ExpectError
	srotg.assign( 1.0, 2.0, out, 1, true ); // $ExpectError
	srotg.assign( 1.0, 2.0, out, 1, false ); // $ExpectError
	srotg.assign( 1.0, 2.0, out, 1, null ); // $ExpectError
	srotg.assign( 1.0, 2.0, out, 1, [] ); // $ExpectError
	srotg.assign( 1.0, 2.0, out, 1, {} ); // $ExpectError
	srotg.assign( 1.0, 2.0, out, 1, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the `assign` method is provided an unsupported number of arguments...
{
	const out = [ 0.0, 0.0, 0.0, 0.0 ];

	srotg.assign(); // $ExpectError
	srotg.assign( 1.0 ); // $ExpectError
	srotg.assign( 1.0, 2.0 ); // $ExpectError
    srotg.assign( 1.0, 2.0, out ); // $ExpectError
    srotg.assign( 1.0, 2.0, out, 1 ); // $ExpectError
    srotg.assign( 1.0, 2.0, out, 1, 0, 1 ); // $ExpectError
}
