/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import dlacpy = require( './index' );


// TESTS //

// The function returns a Float64Array...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy( 'row-major', 'all', 2, 2, A, 2, B, 2 ); // $ExpectType Float64Array
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy( 5, 'all', 2, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( true, 'all', 2, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( false, 'all', 2, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( null, 'all', 2, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( void 0, 'all', 2, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( [], 'all', 2, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( {}, 'all', 2, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( ( x: number ): number => x, 'all', 2, 2, A, 2, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a string...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy( 'row-major', 5, 2, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', true, 2, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', false, 2, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', null, 2, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', void 0, 2, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', [], 2, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', {}, 2, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', ( x: number ): number => x, 2, 2, A, 2, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy( 'row-major', 'all', '5', 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', true, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', false, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', null, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', [], 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', {}, 2, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', ( x: number ): number => x, 2, A, 2, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy( 'row-major', 'all', 5, '2', A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, true, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, false, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, null, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, [], A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, {}, A, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, ( x: number ): number => x, A, 2, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a Float64Array...
{
	const B = new Float64Array( 4 );

	dlacpy( 'row-major', 'all', 5, 2, 5, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, true, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, false, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, null, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, [], 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, {}, 2, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, ( x: number ): number => x, 2, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy( 'row-major', 'all', 5, 2, A, '2', B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, true, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, false, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, null, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, [], B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, {}, B, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, ( x: number ): number => x, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a Float64Array...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlacpy( 'row-major', 'all', 5, 2, A, 2, 5, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, 2, true, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, 2, false, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, 2, null, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, 2, [], 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, 2, {}, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, 2, ( x: number ): number => x, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eighth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy( 'row-major', 'all', 5, 2, A, 2, B, '2' ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, 2, B, true ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, 2, B, false ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, 2, B, null ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, 2, B, [] ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, 2, B, {} ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, 2, B, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy(); // $ExpectError
	dlacpy( 'row-major' ); // $ExpectError
	dlacpy( 'row-major', 'all' ); // $ExpectError
	dlacpy( 'row-major', 'all', 5 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, 2 ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, 2, B ); // $ExpectError
	dlacpy( 'row-major', 'all', 5, 2, A, 2, B, 2, 2 ); // $ExpectError
}

// Attached to main export is an `ndarray` method which returns a Float64Array...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy.ndarray( 'row-major', 'all', 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectType Float64Array
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy.ndarray( 5, 'all', 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( true, 'all', 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( false, 'all', 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( null, 'all', 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( void 0, 'all', 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( [], 'all', 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( {}, 'all', 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( ( x: number ): number => x, 'all', 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a string...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy.ndarray( 'row-major', 5, 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', true, 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', false, 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', null, 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', void 0, 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', [], 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', {}, 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', ( x: number ): number => x, 2, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy.ndarray( 'row-major', 'all', '5', 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', true, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', false, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', null, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', [], 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', {}, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', ( x: number ): number => x, 2, A, 2, 0, B, 2, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy.ndarray( 'row-major', 'all', 5, '2', A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, true, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, false, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, null, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, [], A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, {}, A, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, ( x: number ): number => x, A, 2, 0, B, 2, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a Float64Array...
{
	const B = new Float64Array( 4 );

	dlacpy.ndarray( 'row-major', 'all', 5, 2, 5, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, true, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, false, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, null, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, [], 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, {}, 2, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, ( x: number ): number => x, 2, 0, B, 2, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, '2', 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, true, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, false, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, null, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, [], 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, {}, 0, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, ( x: number ): number => x, 0, B, 2, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, '2', B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, true, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, false, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, null, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, [], B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, {}, B, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, ( x: number ): number => x, B, 2, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eighth argument which is not a Float64Array...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, 5, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, true, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, false, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, null, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, [], 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, {}, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, ( x: number ): number => x, 2, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a ninth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, '2', 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, true, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, false, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, null, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, [], 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, {}, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, ( x: number ): number => x, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a tenth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, 2, '0' ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, 2, true ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, 2, false ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, 2, null ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, 2, [] ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, 2, {} ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, 2, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the `ndarray` method is provided an unsupported number of arguments...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	const B = new Float64Array( 4 );

	dlacpy.ndarray(); // $ExpectError
	dlacpy.ndarray( 'row-major' ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all' ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, 2 ); // $ExpectError
	dlacpy.ndarray( 'row-major', 'all', 5, 2, A, 2, 0, B, 2, 0, 10 ); // $ExpectError
}
