/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import dlaswp = require( './index' );


// TESTS //

// The function returns a Float64Array...
{
	const IPIV = new Int32Array( 3 );
	const A = new Float64Array( 6 );

	dlaswp( 'column-major', 3, A, 2, 0, 2, IPIV, 1 ); // $ExpectType Float64Array
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const IPIV = new Int32Array( 3 );
	const A = new Float64Array( 6 );

	dlaswp( 10, 3, A, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( true, 3, A, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( false, 3, A, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( null, 3, A, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( undefined, 3, A, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( [], 3, A, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( {}, 3, A, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( ( x: number ): number => x, 3, A, 2, 0, 2, IPIV, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a number...
{
	const IPIV = new Int32Array( 3 );
	const A = new Float64Array( 6 );

	dlaswp( 'column-major', '3', A, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'column-major', true, A, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'column-major', false, A, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'column-major', null, A, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'column-major', undefined, A, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'column-major', [ '1' ], A, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'column-major', {}, A, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'column-major', ( x: number ): number => x, A, 2, 0, 2, IPIV, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a Float64Array...
{
	const IPIV = new Int32Array( 3 );

	dlaswp( 'row-major', 3, '10', 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, true, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, false, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, null, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, undefined, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, [ '1' ], 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, {}, 2, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, ( x: number ): number => x, 2, 0, 2, IPIV, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a number...
{
	const IPIV = new Int32Array( 3 );
	const A = new Float64Array( 6 );

	dlaswp( 'column-major', 3, A, '2', 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'column-major', 3, A, true, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'column-major', 3, A, false, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'column-major', 3, A, null, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'column-major', 3, A, undefined, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'column-major', 3, A, [], 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'column-major', 3, A, {}, 0, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'column-major', 3, A, ( x: number ): number => x, 0, 2, IPIV, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a number...
{
	const IPIV = new Int32Array( 3 );
	const A = new Float64Array( 6 );

	dlaswp( 'row-major', 3, A, 2, '0', 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, true, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, false, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, null, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, undefined, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, [], 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, {}, 2, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, ( x: number ): number => x, 2, IPIV, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a number...
{
	const IPIV = new Int32Array( 3 );
	const A = new Float64Array( 6 );

	dlaswp( 'row-major', 3, A, 2, 0, '2', IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, true, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, false, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, null, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, undefined, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, [], IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, {}, IPIV, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, ( x: number ): number => x, IPIV, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not an Int32Array...
{
	const A = new Float64Array( 6 );

	dlaswp( 'row-major', 3, A, 2, 0, 2, 'IPIV', 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, 2, true, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, 2, false, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, 2, null, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, 2, undefined, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, 2, [], 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, 2, {}, 1 ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, 2, ( x: number ): number => x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eighth argument which is not a number...
{
	const IPIV = new Int32Array( 3 );
	const A = new Float64Array( 6 );

	dlaswp( 'row-major', 3, A, 2, 0, 2, IPIV, '1' ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, 2, IPIV, true ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, 2, IPIV, false ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, 2, IPIV, null ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, 2, IPIV, undefined ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, 2, IPIV, [] ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, 2, IPIV, {} ); // $ExpectError
	dlaswp( 'row-major', 3, A, 2, 0, 2, IPIV, ( x: number ): number => x ); // $ExpectError
}
